### PR TITLE
Implement smarter numerical ordering

### LINKED ISSUE #5: Implement smarter numerical ordering, such as E9 < E10
This should be handled better:

```rust
#[sorted]
pub enum Test {
    E9,
    E10,
}
```

```console
error: E10 should sort before E9
 --> src/main.rs:6:5
  |
6 |     E10,
  |     ^^^
```

### PR DESCRIPTION
**Pull request by Juici**
Resolves #5.

### CODE REVIEW

File: src/compare.rs [hunk 1]
```diff
@@ -34,20 +34,34 @@ fn cmp_segment(lhs: &str, rhs: &str, mode: UnderscoreOrder) -> Ordering {
...
+        let left = lhs_atoms.next();
+        let right = rhs_atoms.next();
 
-    // For now: asciibetical ordering.
-    lhs.cmp(&rhs)
-}
+        let (left, right) = match (left, right) {
+            (None, None) => return Ordering::Equal,
+            (None, Some(_)) => return Ordering::Greater,
+            (Some(_), None) => return Ordering::Less,
```
Comment (dtolnay) [inline src/compare.rs]: I think these are backwards. "A" \< "Aa", so "A" should be also less than "A0".

File: src/compare.rs [hunk 2]
```diff
@@ -34,20 +34,34 @@ fn cmp_segment(lhs: &str, rhs: &str, mode: UnderscoreOrder) -> Ordering {
...
+            (Some(left), Some(right)) => (left, right),
+        };
 
-fn count_leading_underscores(segment: &str) -> usize {
-    segment.chars().take_while(|c| *c == '_').count()
+        // Compare underscores with respect to mode.
+        match (&left, &right) {
```
Comment (dtolnay) [inline src/compare.rs]: I would prefer to put a derive(Copy) on Atom and then these `&` wouldn't be needed.

File: src/compare.rs [hunk 3]
```diff
@@ -34,20 +34,34 @@ fn cmp_segment(lhs: &str, rhs: &str, mode: UnderscoreOrder) -> Ordering {
...
+            (Atom::Underscore(l), Atom::Underscore(r)) => match l.cmp(&r) {
+                Ordering::Equal => continue,
+                non_eq => return non_eq,
+            },
+            (Atom::Underscore(_), _) if mode == UnderscoreOrder::First => return Ordering::Greater,
+            (Atom::Underscore(_), _) => return Ordering::Less,
+            (_, Atom::Underscore(_)) if mode == UnderscoreOrder::First => return Ordering::Less,
+            (_, Atom::Underscore(_)) => return Ordering::Greater,
```
Comment (dtolnay) [inline src/compare.rs]: UnderscoreOrder should only be applied to leading underscores. Things like GNU sort and vim's sort like to put underscores first (`under__score < under_score`) but in Rust we often want leading underscores (`__Nonexhaustive`) last as a special case.

File: src/atom.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    /// A sequence of digits.
+    Number(u64),
+    /// A sequence of characters.
+    Chars(&'a str),
+    /// A sequence of underscores.
+    Underscore(usize),
+}
+
+impl<'a> PartialOrd<Atom<'a>> for Atom<'a> {
+    fn partial_cmp(&self, other: &Atom<'a>) -> Option<Ordering> {
```
Comment (dtolnay) [inline src/atom.rs]: This should be written as:

```rust
impl PartialOrd for Atom<'_> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
```

The PartialOrd type parameter is defaulted to the right thing.

File: src/atom.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                        b'0'..=b'9' => self.offset += 1,
+                        _ => break,
+                    }
+                }
+
+                // This is safe since we've verified it's valid utf8.
+                let s = &self.bytes[start..self.offset];
+                let s = unsafe { from_utf8_unchecked(s) };
+
+                let int = match s.parse::<u64>() {
```
Comment (dtolnay) [inline src/atom.rs]: This shouldn't be required, and crashes if a number is big.

```console
error: custom attribute panicked
  --> tests/order.rs:30:1
   |
30 | #[remain::sorted]
   | ^^^^^^^^^^^^^^^^^
   |
   = help: message: internal error: entered unreachable code: expected integer
```

There is a simpler way to compare base 10 numbers without converting to base 2: first by length then by ordinary alphanumeric comparison.

File: src/atom.rs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            _ => {
+                let start = self.offset;
+
+                self.offset += 1;
+                while self.offset < self.bytes.len() {
+                    match self.bytes[self.offset] {
+                        b'_' | b'0'..=b'9' => break,
+                        b => {
+                            // Use a copy of the UTF8_CHAR_WIDTH array from core::str to avoid
+                            // relying on unstable internals.
```
Comment (dtolnay) [inline src/atom.rs]: I don't want to do this. Could you find a way to implement this without copying a utf8 table and without unsafe code?

Comment (Juici) [reply src/atom.rs]: We can always go on the assumption that the bytes are valid and not bother checking them.

**Comment by Juici** [comment]
That should fix the issues.

**Comment by dtolnay** [comment]
Published in 0.2.0.

**Pull request merged**